### PR TITLE
feat(qoi): add QOI image decoder support

### PR DESCRIPTION
**Pull request by mk14lib**
## Summary

Add support for the Quite OK Image (QOI) format as an LVGL image decoder.

The QOI decoder provides:
- Support for both file and variable (in-memory) image sources
- Integration with LVGL's image cache system
- Stride-aligned rendering for hardware display pipelines
- Configurable file size and pixel limits for embedded safety
- RGBA to ARGB8888 color format conversion

### Core Components
- **Decoder implementation**: `src/image/lv_qoi.c` with standard info/open/close callbacks following LVGL decoder pattern
- **Public API**: `lv_qoi_init()` / `lv_qoi_deinit()` in `include/lvgl/image/lv_qoi.h`
- **Reference library**: Unmodified MIT-licensed qoi.h from phoboslab/qoi with LVGL memory hooks
- **Configuration**: `LV_USE_QOI`, `LV_QOI_MAX_FILE_SIZE` (64 MB default), `LV_QOI_MAX_PIXELS` (400M pixels default)
- **Initialization**: `lv_qoi_init()` called in `src/lv_init.c` after libpng, before tjpgd

### Key Features
- Dual-source decoding (file paths via lv_fs and in-memory variables)
- Full integration with LVGL image cache subsystem
- Stride-aligned rendering for GPU-compatible row pitches  
- Endian-independent header parsing for cross-platform compatibility
- Comprehensive bounds checking and validation
- Efficient RGBA→ARGB8888 conversion using 64-bit dual-pixel swap
- Memory allocation through LVGL allocator hooks (QOI_MALLOC/QOI_FREE/QOI_ZEROARR)

### Build Results
- ✅ Static analysis passed: all header checks (public-api, include-paths)
- ✅ Code formatting compliant with LVGL style (astyle)
- ✅ Configuration synchronized across Kconfig, lv_conf_template.h, lv_conf_internal.h
- ✅ No compiler warnings introduced

### Test Results
- ✅ **14 comprehensive unit tests** in `tests/src/test_cases/libs/test_qoi.c`:
  - Decoder registration and info retrieval
  - Valid decoding (RGB, RGBA, linear, sRGB colorspace)
  - Rendering with screenshot comparison
  - Error handling (invalid headers, corrupted data)
  - Edge cases (1×1 pixel images, 512×512 large images)
  - Memory leak testing (100 open/close iterations, 40 render cycles)
  - Multiple format support (RGB vs RGBA comparison)
- ✅ Test assets included: `tests/src/test_assets/test_qoi_*.qoi`
- ✅ All tests pass when `LV_USE_QOI=1`, stubs provided when disabled

### Code Quality
- ✅ Comprehensive Doxygen documentation in headers
- ✅ Usage guide: `docs/src/libs/image_support/qoi.mdx`
- ✅ Example: `examples/libs/qoi/lv_example_qoi_1.c` demonstrating file + variable sources
- ✅ Deprecated header redirect in `src/libs/qoi/lv_qoi.h` with warning message
- ✅ Changelog entry: `docs/src/changelog/CHANGELOG.mdx`

## Review Findings

During comprehensive code review, the following improvements were made:

### Fixes Applied
1. **Cache failure handling**: Returns `LV_RESULT_INVALID` when cache add fails (consistent with libpng pattern)
2. **Safe cleanup**: Sets `dsc->decoded = NULL` before returning invalid to prevent dangling pointers
3. **Pixel limit validation**: Enforces user-configurable `LV_QOI_MAX_PIXELS` in addition to QOI_PIXELS_MAX
4. **Function prototypes**: Added `convert_rgba_to_argb8888` to static prototypes section
5. **CHANGELOG formatting**: Fixed malformed markdown entry

## Compatibility

- ✅ **Existing image formats verified**: PNG, BMP, JPEG, GIF, WebP decoders unaffected
- ✅ **No API breaking changes**: All existing LVGL APIs unchanged
- ✅ **No regressions detected**: Decoder follows established patterns
- ✅ **Backward compatible**: Disabled by default via `LV_USE_QOI = 0`

## Explicit Statement

**The QOI reference library (`src/libs/qoi/qoi.h`) was not modified as part of this contribution.**

Only LVGL integration code was added. The reference library remains in its original form from phoboslab/qoi (MIT licensed), with only an EOF newline added by the LVGL code formatter. Memory allocation hooks are defined before including qoi.h to use LVGL's allocator.

### Potential Risks & Mitigations

| Risk | Impact | Mitigation |
|------|--------|------------|
| Memory allocation for large images | Could exhaust heap | Configurable limits: `LV_QOI_MAX_FILE_SIZE` and `LV_QOI_MAX_PIXELS` prevent DoS via malicious headers |
| Decode performance | Large images may decode slowly | LVGL image caching reduces repeated decodes; acceptable for tested sizes up to 512×512 |
| Malicious QOI files | Could crash decoder | Comprehensive validation: magic bytes, dimension bounds, overflow-safe math, corrupted data handling |

### Risk Level: **LOW**
- Reference library unmodified and well-tested
- Follows established LVGL decoder patterns
- Comprehensive error handling prevents crashes
- Disabled by default (opt-in feature)

## Rollback Strategy

To disable QOI decoder if issues arise:

1. Set `LV_USE_QOI = 0` in lv_conf.h
2. Decoder automatically excluded from build
3. No changes to existing functionality
4. Future removal: Delete QOI files and Kconfig entries

## Checklist

- [x] Documentation updated (`docs/src/libs/image_support/qoi.mdx`)
- [x] Examples added (`examples/libs/qoi/lv_example_qoi_1.c`)
- [x] Tests added (14 test cases in `tests/src/test_cases/libs/test_qoi.c`)
- [x] Kconfig updated (`LV_USE_QOI` and related options)
- [x] `lv_conf_template.h` updated
- [x] Config generators run (`lv_conf_internal.h` synchronized)
- [x] Code formatted (`scripts/code-format.py`)
- [x] Code style followed (LVGL naming conventions, Doxygen comments)
- [x] Static checks passed (header validation)

## Additional Notes

Some unrelated files were formatted by the code formatter during the formatting pass (nanovg, opengles, gltf, stb_image, fragment example). This is expected behavior of the LVGL formatter and unrelated to QOI changes.

---

<a href="https://cubic.dev/pr/lvgl/lvgl/pull/10472?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by kisvegabor** [comment]
Hey,

Thank you for the PR. I wonder what are the benefits of QOI compared to other image formats.

**Comment by mk14lib** [comment]
> Hey,
> 
> Thank you for the PR. I wonder what are the benefits of QOI compared to other image formats.

Hey,

Thanks for the question. Compared with formats such as PNG, QOI provides:

- Lossless compression with comparable file sizes.
- Much faster encoding and decoding.
- A very small and simple implementation—around 300 lines of C.
- Single-pass \(O(n)\) processing with low CPU and memory overhead.
- No external dependencies, which makes it particularly suitable for embedded systems and LVGL.

The main trade-off is that QOI prioritizes speed and simplicity over maximum compression efficiency and broad ecosystem support.

Please go through the QOI repository and the official website for the specification, implementation details, and benchmark results:

- https://github.com/phoboslab/qoi
- https://qoiformat.org/

<citation src="1,4"></citation>

**Comment by kisvegabor** [comment]
Thank you! I've found this benchmark: https://qoiformat.org/benchmark/
It basically says: same file size as PNG but 2-3x faster to decode.  

It'd be amazing to allow starting the decoding on an arbitrary pixel (or at least a line). With this we wouldn't need the large buffer to store the decoded image, but we could decode it line by line.

Do you see any chance that we can modify the decoder (and maybe the encoder too) like this?

File: docs/src/changelog/CHANGELOG.mdx [hunk 1]
```diff
@@ -53,6 +53,7 @@ LVGL v9.5.0 continues expanding MPU capabilities while maintaining the lightness
 
 ##### New Features
 
+- **feat(qoi): add QOI image decoder support**
```
Comment (AndreCostaaa) [inline docs/src/changelog/CHANGELOG.mdx]: This is the v9.5 changelog.  Revert please

File: docs/src/libs/image_support/qoi.mdx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+The **QOI** (Quite OK Image) format is a fast, lossless image compression format
+that encodes and decodes images at high speed with a simple, dependency-free
+implementation.
+
+For a detailed introduction, see: https://github.com/phoboslab/qoi.
+
+If enabled in `lv_conf.h` by setting <ApiLink name="LV_USE_QOI" /> to `1`, LVGL will
+register a new image decoder automatically so QOI files can be used directly as an
+image source.
```
Comment (AndreCostaaa) [inline docs/src/libs/image_support/qoi.mdx]:
```suggestion
When <ApiLink name="LV_USE_QOI" /> is enabled, LVGL will
register a new image decoder automatically so QOI files can be used directly as an
image source.
```

Don't mention lv-conf.h specifically as it's not the only way to configure LVGL

File: examples/libs/qoi/about.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+---
+title: "QOI"
+order: 26
+---
+
+QOI (Quite OK Image) decoder integration. Enable with `LV_USE_QOI` in `lv_conf.h`.
```
Comment (AndreCostaaa) [inline examples/libs/qoi/about.md]:
```suggestion
QOI (Quite OK Image) decoder integration. Enable `LV_USE_QOI` in your configuration
```

**Comment by AndreCostaaa** [comment]
> It'd be amazing to allow starting the decoding on an arbitrary pixel (or at least a line). With this we wouldn't need the large buffer to store the decoded image, but we could decode it line by line.

It seems it's not supported by this implementation: https://github.com/phoboslab/qoi#limitations

File: src/image/lv_qoi.c [hunk 4]
```diff
@@ -0,0 +1,525 @@
...
+ * - Requires the QOI reference library (qoi.h / qoi.c) bundled in the
+ *   LVGL qoi integration.
+ * - Requires LV_USE_QOI to be enabled in lv_conf.h.
+ */
+
+/*********************
+ *      INCLUDES
+ *********************/
+#include "lv_image_decoder_private.h"
+#include "../../lvgl.h"
```
Comment (AndreCostaaa) [inline src/image/lv_qoi.c]:
```suggestion
#include "../lvgl_public.h"
```

File: src/image/lv_qoi.c [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+ *
+ * Call lv_qoi_init() during LVGL initialization to register the decoder.
+ * Once registered, LVGL will automatically use this decoder when loading
+ * QOI images via lv_image_create() or lv_image_set_src().
+ *
+ * ## Dependencies
+ *
+ * - Requires the QOI reference library (qoi.h / qoi.c) bundled in the
+ *   LVGL qoi integration.
+ * - Requires LV_USE_QOI to be enabled in lv_conf.h.
```
Comment (AndreCostaaa) [inline src/image/lv_qoi.c]: This documentation doesn't rely need to be here

File: src/image/lv_qoi.c [hunk 6]
```diff
@@ -0,0 +1,525 @@
...
+ *   on both little-endian and big-endian architectures.
+ * - **Bounds Checking**: All memory accesses are guarded by size
+ *   validation against QOI_HEADER_SIZE, QOI_MIN_FILE_SIZE, and
+ *   LV_QOI_MAX_FILE_SIZE to prevent buffer overruns.
+ * - **Post-Processing Support**: Passes decoded buffers through
+ *   lv_image_decoder_post_process() for color format conversion,
+ *   premultiplied alpha handling, or other LVGL pipeline transforms.
+ * - **Memory Safety**: Overrides QOI library memory hooks (QOI_MALLOC,
+ *   QOI_FREE, QOI_ZEROARR) to use LVGL's allocator, ensuring all
+ *   allocations are tracked within the LVGL memory pool.
```
Comment (AndreCostaaa) [inline src/image/lv_qoi.c]: Not sure how pertinent this whole section is either

File: src/libs/qoi/lv_qoi.h [hunk 7]
```diff
@@ -0,0 +1,2 @@
+#warning Include public headers from the `src` folder is deprecated. This file will be removed soon. To ensure your application keeps working, you can choose to include the new header in <include/lvgl/image/lv_qoi.h> or include the main header file <include/lvgl/lvgl.h>
+#include "../../../include/lvgl/lvgl.h"
```
Comment (AndreCostaaa) [inline src/libs/qoi/lv_qoi.h]: This is only for files that used to exist in v9.5, `lv_qoi.h` has never existed in the src folder so please remove this file

File: tests/src/test_cases/libs/test_qoi.c [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+    /* Also test rendering */
+    lv_obj_clean(lv_screen_active());
+
+    lv_obj_t * img = lv_image_create(lv_screen_active());
+    TEST_ASSERT_NOT_NULL(img);
+
+    lv_image_set_src(img, image_path);
+    lv_obj_align(img, LV_ALIGN_CENTER, 0, 0);
+
+    TEST_ASSERT_EQUAL_SCREENSHOT("libs/qoi_1x1.png");
```
Comment (AndreCostaaa) [inline tests/src/test_cases/libs/test_qoi.c]: A whole screenshot for 1 pixel image feels too much, can you instead simply check that the pixel at the expected position is the color you expect it to be and that the neighbors are white?

Also remove the qoi_1x1.png reference images

**Review by AndreCostaaa — CHANGES_REQUESTED** [review]
Overall it looks good, thank you!

**Comment by mk14lib** [comment]
> Hello!

I addressed the review comments and suggestions. But CI/CD pipeline is stuck during the installation of prerequisites for some checks. But I ran the test-suite locally. All tests PASSED.

Thank you!

**Comment by kisvegabor** [comment]
@mk14lib Thank you for the updates. I wonder what is the recommended way to convert a PNG file to C array (as `lv_image_header_t`).

@AndreCostaaa in v10 we need to banchmark and consolidate the list of supported image formats

**Comment by mk14lib** [comment]
> @mk14lib Thank you for the updates. I wonder what is the recommended way to convert a PNG file to C array (as `lv_image_header_t`).
> 
> @AndreCostaaa in v10 we need to banchmark and consolidate the list of supported image formats

Hello @kisvegabor,

To Convert PNG to C array, here are the official tools:
1. **Online:** https://lvgl.io/tools/imageconverter
2. **Offline:** lvgl/scripts/LVGLImage.py. Execute Command: python3 scripts/LVGLImage.py --ofmt C --cf ARGB8888 input.png -o output.c

The converter produces an [lv_image_dsc_t] descriptor whose .header is the [lv_image_header_t]

```c
const lv_image_dsc_t my_img = {
    .header = {
        .magic = LV_IMAGE_HEADER_MAGIC,      // magic number
        .cf    = LV_COLOR_FORMAT_ARGB8888,   // color format
        .w     = 100,                        // width
        .h     = 50,                         // height
        .flags = 0
    },
    .data_size = 100 * 50 * 4,               // w * h * bytes_per_pixel
    .data = my_img_map                        // raw pixel array
};
```

**Comment by kisvegabor** [comment]
I mean C array with a QOI content. So how `examples/libs/qoi/img_qoi_demo.c` was created?

**Comment by mk14lib** [comment]
> I mean C array with a QOI content. So how `examples/libs/qoi/img_qoi_demo.c` was created?

img_qoi_demo.c was created by 

(1) encoding a source image (.png) into a .qoi file, 

```shell
# Use the official qoi converter, refer: https://github.com/phoboslab/qoi
qoiconv input.png output.qoi
```

then 
(2) embedding the raw .qoi binary bytes into a C array using a Python script, wrapped in an lv_image_dsc_t with .cf = LV_COLOR_FORMAT_RAW. The LVGL QOI decoder reads the embedded QOI header itself, so no lv_image_header_t pixel descriptor is required — though .magic must still be set so LVGL recognizes the descriptor as a variable image source.

Here is the script that I have use to generate the `img_qoi_demo.c`. Please read the comments in the script since they answer some questions that may arise.

```python
#!/usr/bin/env python3
"""
qoi_to_c.py - Convert a QOI binary image file into a C source file
containing an `lv_image_dsc_t` descriptor compatible with LVGL.

Note: The generated `lv_image_dsc_t` descriptor intentionally omits the
`.w`, `.h`, and `.flags` header fields. The QOI format is self-describing —
width/height are encoded in the QOI header bytes, and LVGL's QOI decoder
(lvgl_fork/src/image/lv_qoi.c) parses them directly from the data stream,
overriding `header->w`, `header->h`, and `header->cf` itself during
`decoder_info`. The `.flags` field is unnecessary since QOI is fully decoded
into ARGB8888.

Usage:
    python3 qoi_to_c.py input.qoi [--output out.c] [--name img_foo]
"""

import argparse
import os
import struct
import sys

QOI_MAGIC = b"qoif"
QOI_HEADER_SIZE = 14
BYTES_PER_LINE = 16


class QOIError(Exception):
    """Raised when the input file is not a valid QOI image."""


def parse_args(argv):
    parser = argparse.ArgumentParser(
        description="Convert a QOI binary file into an LVGL C source file."
    )
    parser.add_argument("input", help="Path to the input .qoi file")
    parser.add_argument(
        "--output",
        "-o",
        help="Path to the output .c file (default: derived from input filename)",
    )
    parser.add_argument(
        "--name",
        "-n",
        help="Base variable name (default: derived from input filename)",
    )
    return parser.parse_args(argv)


def derive_base_name(input_path):
    """Derive a C identifier base name from the input filename."""
    stem = os.path.splitext(os.path.basename(input_path))[0]
    # Sanitize to a valid C identifier prefix.
    sanitized = "".join(ch if ch.isalnum() or ch == "_" else "_" for ch in stem)
    if not sanitized:
        sanitized = "img"
    if sanitized[0].isdigit():
        sanitized = "img_" + sanitized
    return sanitized


def default_output_path(input_path, base_name):
    """Default output path: same directory as input, <base_name>.c"""
    directory = os.path.dirname(os.path.abspath(input_path))
    return os.path.join(directory, base_name + ".c")


def read_qoi_header(data):
    """Parse and validate the 14-byte QOI header.

    Returns (width, height, channels, colorspace).
    """
    if len(data) < QOI_HEADER_SIZE:
        raise QOIError("File too small to contain a QOI header")

    magic = data[0:4]
    if magic != QOI_MAGIC:
        raise QOIError(
            "Invalid QOI magic: expected 'qoif', got %r" % magic
        )

    width, height = struct.unpack(">II", data[4:12])
    channels = data[12]
    colorspace = data[13]

    if width == 0 or height == 0:
        raise QOIError("Invalid dimensions: width=%d height=%d" % (width, height))
    if channels not in (3, 4):
        raise QOIError("Invalid channel count: %d (expected 3 or 4)" % channels)
    if colorspace not in (0, 1):
        raise QOIError("Invalid colorspace value: %d (expected 0 or 1)" % colorspace)

    return width, height, channels, colorspace


def format_byte_array(data, bytes_per_line=BYTES_PER_LINE):
    """Format raw bytes as C array initializer lines (0xNN, ...)."""
    lines = []
    for i in range(0, len(data), bytes_per_line):
        chunk = data[i : i + bytes_per_line]
        entries = ", ".join("0x%02x" % b for b in chunk)
        lines.append("    " + entries + ",")
    return lines


def attribute_suffix(base_name):
    """Derive the LV_ATTRIBUTE_IMG_* suffix, stripping a leading 'img_' prefix.

    Matches LVGL convention: 'img_qoi_demo' -> LV_ATTRIBUTE_IMG_QOI_DEMO
    """
    if base_name.startswith("img_"):
        return base_name[4:].upper()
    return base_name.upper()


def generate_c_source(base_name, data):
    """Build the complete C source text.

    The descriptor intentionally omits `.w`, `.h`, and `.flags` because the
    QOI format is self-describing: width/height are encoded in the QOI header
    bytes, and LVGL's QOI decoder parses them directly from the data stream,
    overriding `header->w`, `header->h`, and `header->cf` itself during
    `decoder_info`. The `.flags` field is unnecessary since QOI is fully
    decoded into ARGB8888.
    """
    map_name = base_name + "_map"
    dsc_name = base_name
    attr_suffix = attribute_suffix(base_name)

    lines = []
    lines.append("#include \"../../../lvgl.h\"")
    lines.append("#if LV_USE_QOI && LV_BUILD_EXAMPLES")
    lines.append("")
    lines.append("#ifndef LV_ATTRIBUTE_MEM_ALIGN")
    lines.append("    #define LV_ATTRIBUTE_MEM_ALIGN")
    lines.append("#endif")
    lines.append("")
    lines.append("#ifndef LV_ATTRIBUTE_IMG_%s" % attr_suffix)
    lines.append("    #define LV_ATTRIBUTE_IMG_%s" % attr_suffix)
    lines.append("#endif")
    lines.append("")
    lines.append("static const")
    lines.append(
        "LV_ATTRIBUTE_MEM_ALIGN LV_ATTRIBUTE_LARGE_CONST LV_ATTRIBUTE_IMG_%s" % attr_suffix
    )
    lines.append("uint8_t %s[] = {" % map_name)
    lines.extend(format_byte_array(data))
    lines.append("};")
    lines.append("")
    lines.append("const lv_image_dsc_t %s = {" % dsc_name)
    lines.append("    .header = {")
    lines.append("        .magic = LV_IMAGE_HEADER_MAGIC,")
    lines.append("        .cf = LV_COLOR_FORMAT_RAW,")
    lines.append("    },")
    lines.append("    .data_size = sizeof(%s)," % map_name)
    lines.append("    .data = %s" % map_name)
    lines.append("};")
    lines.append("")
    lines.append("#endif /*LV_USE_QOI && LV_BUILD_EXAMPLES*/")
    lines.append("")
    return "\n".join(lines)


def main(argv=None):
    args = parse_args(argv if argv is not None else sys.argv[1:])

    input_path = args.input
    base_name = args.name if args.name else derive_base_name(input_path)
    output_path = args.output if args.output else default_output_path(input_path, base_name)

    try:
        with open(input_path, "rb") as f:
            data = f.read()
    except OSError as exc:
        print("Error: cannot read '%s': %s" % (input_path, exc), file=sys.stderr)
        return 1

    try:
        width, height, channels, colorspace = read_qoi_header(data)
    except QOIError as exc:
        print("Error: %s" % exc, file=sys.stderr)
        return 1

    source = generate_c_source(base_name, data)

    try:
        with open(output_path, "w", encoding="utf-8") as f:
            f.write(source)
    except OSError as exc:
        print("Error: cannot write '%s': %s" % (output_path, exc), file=sys.stderr)
        return 1

    print(
        "Generated %s (%dx%d, channels=%d, colorspace=%d, %d bytes)"
        % (output_path, width, height, channels, colorspace, len(data))
    )
    return 0


if __name__ == "__main__":
    sys.exit(main())

```

**Comment by mk14lib** [comment]
@kisvegabor 
@AndreCostaaa 

Hello, 
I request your attention towards these changes. Please check out!!!
Thanks~

File: src/image/lv_qoi.c [hunk 9]
```diff
@@ -0,0 +1,479 @@
...
+
+/* Exclude stdio functions from qoi.h as LVGL uses its own VFS (lv_fs) */
+#define QOI_NO_STDIO
+
+/* Override QOI memory & array hooks to safely use LVGL's allocator memory pool */
+#define QOI_MALLOC(sz) lv_malloc(sz)
+#define QOI_FREE(p)    lv_free(p)
+#define QOI_ZEROARR(a) lv_memset((a), 0, sizeof(a))
+
+#define QOI_IMPLEMENTATION
```
Comment (suda-morris) [inline src/image/lv_qoi.c:19]: Do we allow the QOI library provided by external? 🤔

**Comment by AndreCostaaa** [comment]
Hi @mk14lib 

We decided to not move forward with this feature for v9.6
We'll get back to you after the release
